src/urllib3/contrib/emscripten/fetch.py

```diff
@@ -403,6 +534,145 @@ def send_request(request: EmscriptenRequest) -> EmscriptenResponse:
+        if response_js.body is not None:
+            # get a reader from the fetch response
+            body_stream_js = response_js.body.getReader()
+            body = _JSPIReadStream(
+                body_stream_js, timeout, request, response, js_abort_controller
+            )
+    else:
+        # get directly via arraybuffer
+        # n.b. this is another async Javascript call.
+        body = _run_sync_with_timeout(
+            response_js.arrayBuffer(),
+            timeout,
+            js_abort_controller,
+            request=request,
+            response=response,
+        ).to_py()
+    response.body = body
+    return response
+
+
+def _run_sync_with_timeout(
+    promise: Any,
+    timeout: float,
+    js_abort_controller: Any,
+    request: EmscriptenRequest | None,
+    response: EmscriptenResponse | None,
+) -> Any:
+    """await a javascript promise synchronously with a timeout set via the
+       AbortController
+
+    Args:
+        promise (Any): Javascript promise to await
+        timeout (float): Timeout in seconds
+        js_abort_controller (Any): A javascript AbortController object, used on timeout
+        request (EmscriptenRequest | None): The request we're currently handling
+        response (EmscriptenResponse | None): Response we're handling if it exists yet.
+
+    Raises:
+        _TimeoutError: If the request times out
+        _RequestError: If the request raises a Javascript exception
+
+    Returns:
+        _type_: The result of awaiting the promise.
+    """
+    timer_id = None
+    if timeout > 0:
+        timer_id = js.setTimeout(
+            js_abort_controller.abort.bind(js_abort_controller), int(timeout * 1000)
+        )
+    try:
+        from pyodide.ffi import run_sync
+
+        # run_sync here uses WebAssembly Javascript Promise Integration to
+        # suspend python until the Javascript promise resolves.
+        return run_sync(promise)
+    except JsException as err:
+        if err.name == "AbortError":
+            raise _TimeoutError(
+                message="Request timed out", request=request, response=response
+            )
+        else:
+            raise _RequestError(message=err.message, request=request, response=response)
+        # raise (something)
+    finally:
+        if timer_id is not None:
+            js.clearTimeout(timer_id)
+
+
+def has_jspi() -> bool:
+    """Return true if jspi can be used.
+
+    This requires both browser support and also webassembly
+    to be in the correct state - i.e. that the javascript
+    call into python was async not sync."""
+    try:
+        from pyodide.ffi import run_sync  # noqa: F401
+
+        try:
+            from pyodide.ffi import can_run_sync
+        except ImportError:
+            from pyodide_js._module import (  # type: ignore[import-not-found]
+                validSuspender,
+            )
+
+            def can_run_sync() -> bool:
+                return bool(validSuspender.value)
```

Comment:
I think it would make sense to just drop this polyfill now and only support JSPI on Pyodide >= 0.26.2.
```suggestion
        from pyodide.ffi import can_run_sync
```